pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go

```diff
@@ -638,13 +646,84 @@ func (a *index) inRevision(obj any) bool {
 
 // All return all known workloads. Result is un-ordered
 func (a *index) All() []model.AddressInfo {
-	res := slices.Map(a.workloads.List(), modelWorkloadToAddressInfo)
+	var res []model.AddressInfo
+	var localNetwork string
+	if ln := a.networks.LocalSystemNamespace.Get(); ln != nil {
+		localNetwork = *ln
+	} else {
+		log.Warnf("Local system namespace is not set, cannot determine local network for ambient index." +
+			" Defaults to an unset network")
+	}
+	// Add all workloads
+	for _, wl := range a.workloads.List() {
+		// If EnableAmbientMultiNetwork is enabled, we only want to add workloads that are local to this cluster
+		// Non cluster local workloads will be added by the service lookup
+		if features.EnableAmbientMultiNetwork && wl.Workload.Network != localNetwork {
+			continue
+		}
+		res = append(res, wl.AsAddress)
+	}
+	// Add all services
 	for _, s := range a.services.List() {
+		if features.EnableAmbientMultiNetwork && s.Scope == model.Global {
+			// If EnableAmbientMultiNetwork is enabled, we want to add workloads that correspond to global services
+			for _, wl := range a.workloads.ByServiceKey.Lookup(s.ResourceName()) {
+				if wl.Workload.Network != localNetwork {
+					res = append(res, wl.AsAddress)
+				}
+			}
+		}
 		res = append(res, s.AsAddress)
 	}
 	return res
 }
 
+// AllLocalNetworkGlobalServices returns all known globally scoped services and local
+// waypoints containing global services. Result is un-ordered
+func (a *index) AllLocalNetworkGlobalServices(key model.WaypointKey) []model.ServiceInfo {
+	var res []model.ServiceInfo
+	for _, svc := range a.services.List() {
+		workloads := a.workloads.ByServiceKey.Lookup(svc.ResourceName())
+		// All workloads for a service should belong in the same network
+		// Don't add a service if it is in a different network
+		if len(workloads) != 0 && workloads[0].Workload.Network != key.Network {
+			log.Debugf("Skipping service %s/%s in network %s, as it is not in the network %s"+
+				" of the network gateway", svc.Service.Namespace, svc.Service.Name, key.Network)
+			continue
+		}
+		if svc.Scope != model.Global {
+			// Check if the service is a waypoint. If the service is not a waypoint
+			// or it's a waypoint containing no services then the Lookup will return
+			// an empty list
+			wpSvcs := a.services.ByOwningWaypointHostname.Lookup(NamespaceHostname{
```

Comment:
We confirmed that waypoint services show up here?

Reply:
I'll test this again manually

Reply:
@keithmattix Added a hold until I push a feature flag check around calling `AllLocalNetworkGlobalServices`